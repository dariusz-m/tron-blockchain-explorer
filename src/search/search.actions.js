export const SearchActionTypes = {
    SEARCH: "SEARCH",
    NOTHING_FOUND: "NOTHING_FOUND",
    FOUND_SOMETHING: "FOUND_SOMETHING",
};

export const SearchStatus = {
    SEARCH_ENGINE_HAS_NOT_BEEN_USED: "SEARCH_ENGINE_HAS_NOT_BEEN_USED",
    PENDING: "PENDING",
    FOUND_SOMETHING: "FOUND_SOMETHING",
    NOTHING_WAS_FOUND: "NOTHING_WAS_FOUND",
};

export const search = (value) => {
    return {
        type: SearchActionTypes.SEARCH,
        payload: value
    };
};

export const nothingFound = () => {
    return {
        type: SearchActionTypes.NOTHING_FOUND,
    };
};

export const foundSomething = () => {
    return {
        type: SearchActionTypes.FOUND_SOMETHING,
    };
};
