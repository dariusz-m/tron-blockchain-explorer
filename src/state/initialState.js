const initialState = {
    router: {"location": null},
    tokens: [],
    selectedBlock: {number: -1, witnessId: -1, hash: "", parentHash: ""}
};

export default initialState;
