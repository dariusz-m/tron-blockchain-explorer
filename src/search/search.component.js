import React from "react";
import {connect} from "react-redux";
import {bindActionCreators} from "redux";

import {search} from "./search.actions";
import PropTypes from "prop-types";

export class Search extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            value: ''
        };

        this.onChange = this.onChange.bind(this);
        this.search = this.search.bind(this);
    }

    search(event) {
        event.preventDefault();
        event.stopPropagation();
        this.props.actions.search(this.state.value);
    }

    onChange(event) {
        this.setState(() => {
            return {value: event.target.value};
        });
        event.persist();
        event.preventDefault();
    }

    render() {
        return (
            <div className="wrapper-search">
                <a className="logo"/>
                <form className="search-widget" onSubmit={this.search}>
                    <input className="search-bar"
                           id="search-bar"
                           placeholder="Transaction, block, account, token..."
                           autoFocus
                           type="text"
                           value={this.state.value}
                           onChange={this.onChange}
                    />
                        <a className="btn search" id="btn-search" onClick={this.search}>
                            &nbsp;
                        </a>
                </form>
            </div>
        );
    }
}

Search.propTypes = {
    actions: PropTypes.shape({
        search: PropTypes.func.isRequired,
    }),
};


function mapDispatchToProps(dispatch) {
    return {
        actions: bindActionCreators({
            search: search,
        }, dispatch)
    };
}

export default connect(null, mapDispatchToProps)(Search);
