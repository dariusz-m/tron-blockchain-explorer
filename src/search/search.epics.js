import { of, merge } from 'rxjs';
import { mergeMap, catchError, pluck, bufferCount} from 'rxjs/operators';
import { ajax } from 'rxjs/ajax';

import {foundSomething, nothingFound} from './search.actions';
import {accountDetailsHaveBeenLoaded} from '../account-summary/account-summary.actions';
import {SearchActionTypes} from "./search.actions";
import {push} from "react-router-redux";
import {blockHasBeenLoaded} from "../block/block.actions";
import {TRON_ORGIN} from "../config/tron-api.config";

export const TYPE_OF_REQUESTS = {
    BLOCK_BY_NUMBER: "BLOCK_BY_NUMBER",
    BLOCK_BY_HASH: "BLOCK_BY_HASH",
    TRANSACTION_BY_HASH: "TRANSACTION_BY_HASH",
    TOKEN_BY_NAME: "TOKEN_BY_NAME",
    ACCOUNT_BY_ADDRESS: "ACCOUNT_BY_ADDRESS",
};

export const LOAD_DATA_ACTIONS = {
    [TYPE_OF_REQUESTS.BLOCK_BY_NUMBER]: blockHasBeenLoaded,
    [TYPE_OF_REQUESTS.BLOCK_BY_HASH]: blockHasBeenLoaded,
    [TYPE_OF_REQUESTS.TRANSACTION_BY_HASH]: () => {},
    [TYPE_OF_REQUESTS.TOKEN_BY_NAME]: () => {},
    [TYPE_OF_REQUESTS.ACCOUNT_BY_ADDRESS]: accountDetailsHaveBeenLoaded,
};

export const REDIRECT_ACTIONS = {
    [TYPE_OF_REQUESTS.BLOCK_BY_NUMBER]: (block) => push(`/block/${block.number}`),
    [TYPE_OF_REQUESTS.BLOCK_BY_HASH]: (block) => push(`/block/${block.hash}`),
    [TYPE_OF_REQUESTS.TRANSACTION_BY_HASH]: (transaction) => push(`/transaction/${transaction.hash}`),
    [TYPE_OF_REQUESTS.TOKEN_BY_NAME]: (token) => push(`/token/${token.name}`),
    [TYPE_OF_REQUESTS.ACCOUNT_BY_ADDRESS]: (account) => push(`/account/${account.address}`),
};

export const makeRequest = (url, typeOfRequest) =>
    ajax(url).pipe(
        pluck('response', 'data'),
        mergeMap((data) => of({type: typeOfRequest, data: data, error: ''})),
        catchError((error) => of({type: typeOfRequest, data: [], error: error})),
    );

export const search = action$ => {
    return action$.ofType(SearchActionTypes.SEARCH)
        .pipe(
            mergeMap((action) => merge(
                makeRequest(
                    `${TRON_ORGIN}/api/block?sort=-number&limit=1&number=${action.payload}`,
                    TYPE_OF_REQUESTS.BLOCK_BY_NUMBER
                ),
                makeRequest(
                    `${TRON_ORGIN}/api/block?sort=-number&limit=1&hash=${action.payload}`,
                    TYPE_OF_REQUESTS.BLOCK_BY_HASH
                ),
                makeRequest(
                    `${TRON_ORGIN}/api/account?sort=-balance&limit=1&address=${action.payload}`,
                    TYPE_OF_REQUESTS.ACCOUNT_BY_ADDRESS
                )
            ).pipe(
                bufferCount(3),
                mergeMap((responses) => {
                    for(let i=0; i < responses.length; i++) {
                        if (responses[i].data.length > 0) {
                            const loadData = LOAD_DATA_ACTIONS[responses[i].type];
                            const redirect = REDIRECT_ACTIONS[responses[i].type];
                            return of(
                                loadData(responses[i].data[0]),
                                foundSomething(),
                                redirect(responses[i].data[0]),
                            );
                        }
                    }
                    return of(nothingFound());
                }),
                catchError((error) => {
                    console.error(error);
                    return of(nothingFound());
                }),
            )),
        );
};
