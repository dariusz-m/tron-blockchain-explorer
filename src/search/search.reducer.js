import initialState from '../state/initialState';
import {SearchActionTypes, SearchStatus} from "./search.actions";

export default (state = initialState.searchEngine, action) => {
    switch (action.type) {
        case SearchActionTypes.NOTHING_FOUND:
            return {...state, status: SearchStatus.NOTHING_WAS_FOUND};
        case SearchActionTypes.FOUND_SOMETHING:
            return {...state, status: SearchStatus.FOUND_SOMETHING};
        default:
            return state;
    }
};
