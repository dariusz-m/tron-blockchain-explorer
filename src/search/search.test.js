import React from 'react';
import renderer from 'react-test-renderer';

import configureStore from "../state";
import {Search} from "./search.component";
import {search, SearchStatus} from "./search.actions";
import initialState from "../state/initialState";
import {mockSearchRequests} from "./search.epics.test";


describe('Search tests', () => {
    let store;

    beforeEach(() => {
        store = configureStore();
    });

    it('Renders correctly', () => {
        const actions = {
            search: jest.fn(),
        };

        const tree = renderer.create(
            <Search
                actions={actions}
            />
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('Search engine should not find anything', (done) => {
        expect.hasAssertions();
        const searchTerm = "dummyValue";
        mockSearchRequests({searchTerm: searchTerm});
        const expectedState = {
            ...initialState,
            searchEngine: {
                ...initialState.searchEngine,
                status: SearchStatus.NOTHING_WAS_FOUND
            }
        };

        store.dispatch(search(searchTerm));

        store.subscribe(() => {
            expect(store.getState()).toEqual(expectedState);
            done();
        });
    });
});

