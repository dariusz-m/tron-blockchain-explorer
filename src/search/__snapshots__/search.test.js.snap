// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search tests Renders correctly 1`] = `
<div>
  <form
    className="search"
  >
    <input
      onChange={[Function]}
      type="text"
      value=""
    />
    <button
      onClick={[Function]}
    >
      Search
    </button>
  </form>
</div>
`;
